src/core/server/saved_objects/validation/validator.ts

```diff
@@ -0,0 +1,129 @@
+ *
+ * See {@link SavedObjectsValidationFunction} for custom validation.
+ *
+ * @public
+ */
+export type SavedObjectsValidationSpec<A extends SavedObjectAttributes> =
+  | ObjectType
+  | SavedObjectsValidationFunction<A>;
+
+/**
+ * A map of {@link SavedObjectsValidationSpec | validation specs} to be used for a given type.
+ * The map's keys must be valid semver versions.
+ *
+ * Any time you change the schema of a {@link SavedObjectsType}, you should add a new entry
+ * to this map for the Kibana version the change was introduced in.
+ *
+ * @example
+ * ```typescript
+ * const validationMap: SavedObjectValidationMap = {
+ *   '1.0.0': schema.object({
+ *     foo: schema.string(),
+ *   }),
+ *   '1.1.0': schema.object({
+ *     foo: schema.oneOf([schema.string(), schema.boolean()]),
+ *   }),
+ *   '2.1.0': (data) => {
+ *     if (typeof data.bar !== 'string') {
+ *       throw new Error(`[bar]: expected value of type [string] but got [${typeof data.bar}]`);
+ *     }
+ *     if (typeof data.foo !== 'string' && typeof data.foo !== 'boolean') {
+ *       throw new Error(`[foo]: expected value of type [string,boolean] but got [${typeof data.foo}]`);
+ *     }
+ *   }
+ * }
+ * ```
+ *
+ * @public
+ */
+export interface SavedObjectsValidationMap<
+  A extends SavedObjectAttributes = SavedObjectAttributes
+> {
+  [version: string]: SavedObjectsValidationSpec<A>;
+}
+
+/**
+ * Helper class that takes a {@link SavedObjectsValidationMap} and runs validations for a
+ * given type based on the provided Kibana version.
+ *
+ * @internal
+ */
+export class SavedObjectsTypeValidator<A extends SavedObjectAttributes = SavedObjectAttributes> {
+  private readonly type: string;
+  private readonly validationMap: SavedObjectsValidationMap<A>;
+
+  constructor({
+    type,
+    validationMap,
+  }: {
+    type: string;
+    validationMap: SavedObjectsValidationMap<A> | (() => SavedObjectsValidationMap<A>);
+  }) {
+    this.type = type;
+    this.validationMap = typeof validationMap === 'function' ? validationMap() : validationMap;
+  }
+
+  public validate(kibanaVersion: string, data: A): void {
+    const validationRule = this.validationMap[kibanaVersion];
+    if (!validationRule) {
+      return; // no matching validation rule could be found; proceed without validating
+    }
+
+    if (isConfigSchema(validationRule)) {
+      validationRule.validate(data, {});
+    } else if (isValidationFunction(validationRule)) {
+      this.validateFunction(data, validationRule);
+    } else {
+      throw new ValidationError(
+        new SavedObjectsValidationError(
+          `The ${kibanaVersion} validation for saved object of type [${this.type}] is malformed.`
+        )
+      );
+    }
+  }
+
+  private validateFunction(data: A, validateFn: SavedObjectsValidationFunction) {
+    try {
+      validateFn({ ...data }); // shallow clone to prevent mutation
+    } catch (err) {
+      throw new ValidationError(new SavedObjectsValidationError(err));
+    }
```

Comment:
I borrowed from the implementation in the http service here so that the validator would throw a proper `ValidationError`, which requires the same hack of setting the prototype explicitly which is included in `validator_error.ts`.

But in the SO repository, I ended up just catching errors and throwing them as a SO `badRequestError` anyway. I kept the `ValidationError` here so that the errors from the validator class are consistent (whether you are using config schemas or a custom function, any errors thrown will be `ValidationErrors`). But in reality this doesn't really do anything for end users, so I'm also happy to remove and just throw plain errors if anyone feels strongly about it.

Reply:
Now that it's here, I'd say it's fine to keep them. It's a good practice anyway overall, and may come handy if we change the implementation somehow.